### DISCUSSION TITLE
Companion vs xDrip+ Sync Follower, Data Source On Followers

### DISCUSSION THREAD

**Discussion by jcisml-G59** [NightscoutFoundation/xDrip#3661] [General]
I have several devices to read the Dexcom G7 as follows:
1. Main phone with the G7 App and xDrip+ set as Companion.
2. All other smart devices (phones to follow the main: xDrip+ set as xDrip+ Companion.

Recently, I happened to see a new phone (transferred from the main but used as xDrip+ Sync Follower with NO Dexcom G7 app but set as Companion) still work seamlessly. For curiosity as I wondered, I changed all other phones with xDrip+ set as xDrip+ Sync Follower to xDrip+ Companion and see them working continuously with NO difference. This contradicts my understanding about Companion setting as the Dexcom G7 App is not installed on any of these devices. The only source is xDrip+ fetching the readings from xDrip+ the main phone in #1. However, Samsung S8 Ultra tab stopped fetching the readings soon after Data Source in xDrip+ was changed to Companion from xDrip+ Sync Follower. Other phones continuously working set as Companion are Samsung S22 Ultra, Flip 6 all running the latest Android.
These two pictures show the S22 Ultra running xDrip+ set as Companion without Dexcom G7 app, fetching the readings from the main phone.
![xDrip+DataSource](https://github.com/user-attachments/assets/5ea67078-f4f7-478c-afa9-933a3e7d1fe2)
![xDrip+](https://github.com/user-attachments/assets/74b0a8e8-ddb0-4086-aac8-a56f115ca1d7)

**Comment by Navid200** [comment]
I'm sorry it is hard to read your description and understand.  

If you have readings in xDrip, you can transfer them to any other xDrip on any other phone through xDrip sync.
xDrip sync is a follower scheme.
So, you should not be surprised if you have a phone that follows your main phone if it does not have Dexcom app on it.

**Reply by jcisml-G59** [reply]
Sorry for the confusion. My question is what the difference is between "Companion" & "xDrip+ Sync Follower" as Data Source on following phones because I recently see all the following phones still working set as "Companion" as set as "xDrip+ Sync Follower". My understand was Companion is only for the main phone with Dexcom G7 app also installed as the main source while xDrip+ Sync Follower is for the following phones without Dexcom G7 app installed working more like Dexcom Share Follow. For reasonable curiosity, I just wonder why other following phones still fetch the readings even when they are set as "Companion" without Dexcom G7 app installed.

**Reply by Navid200** [reply]
You cannot set xDrip to be a sync follower and also to get readings from a companion app.
Just go to the hardware data source page and you will see it is either set to Companion app or xDrip sync.  It cannot be both.

If that is not clear, please ask me to explain.
If that is clear, please ask your question again.

**Reply by PhilTU** [reply]
this is the matter I raised in https://github.com/NightscoutFoundation/xDrip/discussions/3658

**Reply by Navid200** [reply]
At this point, I see no issue.

**Reply by PhilTU** [reply]
the two images are from a follower phone which doesn't have the official app which the OP says is receiving data from a master phone which does. The phone in the images was previously an xDrip+ sync follower. It is now configured (inappropriately) as a companion app data source.

(I hope I got that right)

**Reply by jcisml-G59** [reply]
Yes, you got it down to the point. The phone was set as xDrip+ Sync Follower and now configured as Companion  with no Dexcom G7 app installed but still getting the reading from the Master phone. That is the question I raised.

**Reply by Navid200** [reply]
@jcisml-G59 Do you see 124 in green in the notification area?
What app is that notification from?

**Reply by jcisml-G59** [reply]
It is from xDrip+ set as Companion WITHOUT Dexcom G7 app, getting the reading (124) from the main phone xDrip+ along with Dexcom G7 paired to the G7. I am wondering how it works set as Companion instead of xDrip+ Sync Follower WITHOUT Dexcom G7.
Samsung S8 Ultra Tab does not get the readings if it is set as Companion, only getting the readings if set as xDrip+ Sync Follower.

**Reply by Navid200** [reply]
xDrip Number icon in notification area?
That is shown in white not green.  How did you change it to green?

**Reply by jcisml-G59** [reply]
It is set in Display settings, I remember.

**Comment by PhilTU** [comment]
Can you share the System status page - classic or identify the release version

**Reply by Navid200** [reply]
I hope to first establish an understanding that the hardware data source can either be set to xDrip sync or Companion app.  It cannot be set to both.  If that is not understood, looking at the status page could actually cause more confusion.

**Reply by PhilTU** [reply]
yes. Doesn't help when there are multiple devices.

I was interested in verison as I have only tested on latest nightly. I did observe that once switched to companion (from xDrip+ sync follower) it continued to recieve data but didn't display it. I wondered if an older version might behave differently. Just trying to ratioalise what the OP observed.

**Reply by Navid200** [reply]
If you change the hardware data source from xDrip sync to anything else and leave the master as is, the master will keep sending data to the follower phone.  The log you see that says we have received data even though we are not a follower is related to that.  It has nothing to do with Companion app.

**Reply by PhilTU** [reply]
Indeed. So why is the OP seeing data on a phone they have switched to comanion app, that is the question.

There should be no data collected by a phone set to companion app that has no vendor app, We're all agreed on that.

Companion app and xDrip+ sync follower are entirely different data source concepts. I couldn't explain why they saw what they did.

**Reply by jcisml-G59** [reply]
Latest 9/4 installed but the question arose way from the previous release version.
![xDripStatus](https://github.com/user-attachments/assets/f39e2f33-8e2b-4f0b-a51e-c453fb623a6a)

**Reply by jcisml-G59** [reply]
Of course, only one Data Source can only be set. I wonder I might not have made myself clear enough to Navid200.

**Reply by Navid200** [reply]
What app is the green notification from?

**Reply by jcisml-G59** [reply]
Yes, that is why a question arose, What is the difference between Companion and xDrip+ Sync Follower on the following phone. Also, why phones get the readings while tablet doe not. It is not a big deal as long as it works but just stimulates my curiosity as technically it does not make sense. As PhilTU explained, it was also my understanding that Companion requires the official Dexcom G7 app while xDrip+ Sync Follower does not, just follows the Master. Well, curiosity killed the cat, I guess.

**Reply by Navid200** [reply]
You are running an app that created the green numbers in the notification area.  It cannot be xDrip.
I also want to find an answer.  Can you answer please?

**Reply by jcisml-G59** [reply]
As I mentioned, glucodatahandler is just a handler for watch complications using xDrip+ as Data Source. As far as I know, only apps which can get readings from the Dexcom G7 are either the official Dexcom G7 app and/or xDrip+. All other smart watch related apps like Blose, G-Watch Wear have Data Source set with either Dexcom Share Follow or xDrip+. The same is true for glucodatahandler. It stops reading when xDrip+ stops reading.

**Reply by Navid200** [reply]
You are responding to an old message.  Please see my other request.  Please work with me.  I am not interested in proving you wrong.  I am looking for an explanation.  I think we have a common goal.

**Reply by PhilTU** [reply]
> What app is the green notification from?

Glucodatahandler, which collects from xDrip either via Local Broadcast or Broadcast service API for feeding to smart watches.

So the green notification is a reflection of the xDrip+ glucose value. I have tried to reproduce the original with and without the Glucodatahandler present and failed in all cases. It isn't a vendor app glucose.

**Comment by PhilTU** [comment]
Is this notification from xDrip+ ? 
![image](https://github.com/user-attachments/assets/daba2f4e-7cc7-4c8d-b230-24acf7037363)

**Reply by jcisml-G59** [reply]
That is from glucodatahandler with xDrip+ set as Data Source, so technically yes.

**Reply by Navid200** [reply]
Enable data tables.
What does BG data table show?
https://navid200.github.io/xDrip/docs/Datatables.html

**Reply by Navid200** [reply]
Will readings stop if you only disable Bluetooth?
Will readings stop if you only disable WiFi and cell?

**Reply by jcisml-G59** [reply]
Here is the BG Data Table as of now.
![xDripDataTable](https://github.com/user-attachments/assets/b53659fb-124b-4845-820a-77ad2c3ca1c8)

**Reply by jcisml-G59** [reply]
BT off: Reading continues with no letup.
Wifi off (no cell data): Reading stops.

**Reply by Navid200** [reply]
That's odd.

With WiFi and cell enabled and readings resumed, go to the xDrip sync settings page.  Take note of the last character of "Handset Group Security Sync key".
Change the last character to something else and OK.

Do readings stop?

**Reply by jcisml-G59** [reply]
From all the comments and my observation, it seems:

1. with current settings originally set with xDrip+ Sync Follower as Data Source, changed to Companion, it works either way. (My conjecture is it still works as xDrip+ Sync Follower no matter which is set as Data Source, in which case Companion is overridden by xDrip+ Sync Follower settings.)

Now, I understand a bit better about it with many thanks. Special thanks to both of PhilTU and Navid200.

**Reply by Navid200** [reply]
xDrip is open source.
If this is a bug, we will need to fix it.
If you help us, we can improve xDrip.

But, it is up to you.
This is not intended behavior.  You can consider that you have learnt what it does and even tell people about it.  I hope you won't.
This could be a bug and we can fix it if you help us figure out what is going on.

**Reply by jcisml-G59** [reply]
Of course, I also would consider it as a bug as it contradicts what they are designed for in behavior, probably a minor bug for an easy fix. I started this discussion only for my curiosity as a QA professional who had always been eager to see more diabetic people use and get the best of xDrip+ in their diabetes management. I have always been trying to spread a lot about superior functionality of xDrip+ in Dexcom subreddit for over a year and feel sorry that most there are not tech-savvy and even fail to get the best out of the G7 which has never failed me since I switched over from the G6.
I thought I had provided all the info regarding this issue but am always happy to provide more upon your request.

**Reply by Navid200** [reply]
Can you enable WiFi so that you have readings again on the phone with no Dexcom app on and set to Companion app?
Then, go to Sync settings and change the last character of "Handset Group Security Sync key".
Does it stop providing readings?

**Reply by jcisml-G59** [reply]
That is exactly what I suspected when I wrote my sort of conclusion. I just did and wait for the next reading time to confirm. I feel in my bones that it is causing the issue because those devices were originally set as xDrip+ Sync Follower with the same Sync Key. 
But, as you stated, it is still considered a BUG as the Sync Key must have been obsolete or sent behind the scene as soon as it is changed to Companion mode. Will report the result in about 5 min.

**Reply by jcisml-G59** [reply]
Just confirmed it seems to be the culprit. No readings.   However, another thing bugging me is why table works as intended (in other words, it stops reading if changed to Companion). All the same latest Android 14 (Samsung S9 Ultra Tab).
![xDripSyncKey](https://github.com/user-attachments/assets/6806c49f-0672-4604-9800-2b5c6b90865f)

**Reply by Navid200** [reply]
Sorry, I don't follow.
You changed the sync key and it stopped working?  

I will do everything I can to answer your questions.  But, would you please be clear?

**Reply by jcisml-G59** [reply]
Yes, I did change the last digit of the Sync Key to see it stopped as the picture shows. Changed it back to what it was to see it starts working again as normal. Summary:

1. Phones: Original Sync Follower setting overrides Companion setting in Data Source.
2. Tablet: If Data Source is changed to Companion, it stops reading.

**Reply by Navid200** [reply]
On the phone that works with hardware data source set to companion app with no Dexcom app installed, please post a screenshot of the classic status page.

**Reply by PhilTU** [reply]
Just to say I added glucodatahandler into the mix on my test phones and fed it data two different ways. I couldn't reproduce the issue.

When switched to Companion app data stops on phone with no vendor app. Revert to xDrip sync following and it starts again. Sync settings are preserved.

Context - I set up a pair of phones on latest nightly from scratch. Neither had a history of using either Sync or Companion app.

**Reply by jcisml-G59** [reply]
It sure is strange as it happens only on two of my Samsung phones (S22 Ultra & Flip 6). However, Tab S8 Ultra behaves as your test phones behave, stopping in Companion mode.

**Comment by jcisml-G59** [comment]
Does it do it?
![xDripStatus](https://github.com/user-attachments/assets/63f26d6c-aa4b-4fb2-a979-b3b31e4a1ac5)

**Reply by Navid200** [reply]
Yes, thanks.
I should be able to recreate it.
Thanks

**Reply by Navid200** [reply]
No, I cannot recreate the problem.
As expected, xDrip refuses to get readings from any source other than the one chosen by hardware data source.

I don't know why you experience this.

**Reply by Navid200** [reply]
I get a log saying "Received remote BG packet but we are not set as follower" and readings are not updated.  So, I get no new readings after changing the data source from sync to companion app.

**Reply by jcisml-G59** [reply]
Well, I just noted the newest version (09.17) and updated with it. Alas, now the newest version seems to have disabled Companion on my main phone with the Dexcom G7 app. The only way to get it working as it used to seems to be downgrade it to the one that it was working (09.04). What gives?

Edit: Just reinstalled the old version (09.04) but still does not work for some reason. What should I do to correct it?

**Reply by Navid200** [reply]
There has been no change to the Companion App mode.

I am running the latest Nightly, 2024.09.17.  I just installed Dexcom app and set up companion app mode and it is working fine.

Are you sure you are getting Dexcom app notifications in your notification area?

**Reply by jcisml-G59** [reply]
Well, nothing has changed except xDrip+ was updated with the newest version, 09.17. Should I uninstall and reinstall from scratch?

**Reply by Navid200** [reply]
No, you should verify that you can see Dexcom app notification if you pull down from the top.  I already asked you that.

**Reply by jcisml-G59** [reply]
![DexNoti](https://github.com/user-attachments/assets/611ab5f2-5079-4ea4-8b77-56eeb7f755c4)

**Reply by Navid200** [reply]
Have you enabled Companion Bluetooth?  Please do if not.
Please restart the phone.

**Reply by jcisml-G59** [reply]
1. Have never heard of Companion Bluetooth as I have never done. Where do I see the option to enable it?
2. Already restarted the phone.

**Reply by Navid200** [reply]
It's at `Settings` &#8722;> `Less common settings` &#8722;> `Bluetooth Settings` &#8722;> `Companion Bluetooth`

**Reply by jcisml-G59** [reply]
Believe it or not, I see it already enabled.

**Reply by Navid200** [reply]
Can you tap on restart collector on the classic status page?

**Reply by Navid200** [reply]
If installing the xDrip version of 2024.09.04 works and installing the 2024.09.17 fails, would you please install the following in sequence and tell us the very first one that fails?

https://github.com/NightscoutFoundation/xDrip/releases/tag/2024.09.04
https://github.com/NightscoutFoundation/xDrip/releases/tag/2024.09.15
https://github.com/NightscoutFoundation/xDrip/releases/tag/2024.09.15b
https://github.com/NightscoutFoundation/xDrip/releases/tag/2024.09.15c
https://github.com/NightscoutFoundation/xDrip/releases/tag/2024.09.16
https://github.com/NightscoutFoundation/xDrip/releases/tag/2024.09.17
